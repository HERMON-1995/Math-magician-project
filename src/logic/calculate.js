import operate from './operate';

function isNumber(item) {
  return /\d/.test(item);
}

export default function calculate(obj, buttonName) {
  const { total, next, operation } = obj;

  switch (buttonName) {
    case 'AC':
      return {
        total: null,
        next: null,
        operation: null,
      };
    case '+/-':
      return {
        ...obj,
        next: (next ? -1 * parseFloat(next) : null),
        total: (total ? -1 * parseFloat(total) : null),
      };
    case '.':
      if (next && next.includes('.')) {
        return obj;
      }
      return {
        ...obj,
        next: (next ? `${next}.` : '0.'),
      };
    case '=':
      if (next && operation) {
        return {
          total: operate(total, next, operation),
          next: null,
          operation: null,
        };
      }
      return {};
    default:
      if (isNumber(buttonName)) {
        if (buttonName === '0' && next === '0') {
          return {};
        }
        if (operation) {
          return {
            ...obj,
            next: (next && next !== '0') ? next + buttonName : buttonName,
          };
        }
        return {
          ...obj,
          next: (next && next !== '0') ? next + buttonName : buttonName,
          total: null,
        };
      }
      if (!next && total && !operation) {
        return {
          ...obj,
          operation: buttonName,
        };
      }
      if (operation) {
        if (total && !next) {
          return {
            ...obj,
            operation: buttonName,
          };
        }
        return {
          total: operate(total, next, operation),
          next: null,
          operation: buttonName,
        };
      }
      return {
        ...obj,
        total: next,
        next: null,
        operation: buttonName,
      };
  }
}
