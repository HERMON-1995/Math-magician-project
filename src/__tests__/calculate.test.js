import calculate from '../logic/calculate';

describe('test on calculate.js function', () => {
    test('when AC button is pressed', () => {
      const output = calculate({
        total: '8',
        next: '6',
        operation: '',
      }, 'AC');
      expect(output).toEqual({
        total: null,
        next: null,
        operation: null,
      });
    });
    test('= button', () => {
      const output = calculate({
        total: '5',
        next: '5',
        operation: '+',
      }, '=');
      expect(output.total).toBe('10');
    });
  });
