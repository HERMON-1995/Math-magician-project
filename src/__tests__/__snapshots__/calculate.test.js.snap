// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="calculator"
  >
    <h1
      class="title"
    >
      ** Calculator **
    </h1>
    <div
      class="display-section"
    >
      <p
        class="display-value"
      >
          
      </p>
    </div>
    <button
      type="button"
    >
      AC
    </button>
    <button
      type="button"
    >
      +/-
    </button>
    <button
      type="button"
    >
      %
    </button>
    <button
      class="operator"
      type="button"
    >
      รท
    </button>
    <button
      type="button"
    >
      7
    </button>
    <button
      type="button"
    >
      8
    </button>
    <button
      type="button"
    >
      9
    </button>
    <button
      class="operator"
      type="button"
    >
      x
    </button>
    <button
      type="button"
    >
      4
    </button>
    <button
      type="button"
    >
      5
    </button>
    <button
      type="button"
    >
      6
    </button>
    <button
      class="operator"
      type="button"
    >
      -
    </button>
    <button
      type="button"
    >
      1
    </button>
    <button
      type="button"
    >
      2
    </button>
    <button
      type="button"
    >
      3
    </button>
    <button
      class="operator"
      type="button"
    >
      +
    </button>
    <button
      class="zero"
      type="button"
    >
      0
    </button>
    <button
      type="button"
    >
      .
    </button>
    <button
      class="operator"
      type="button"
    >
      =
    </button>
  </div>
</DocumentFragment>
`;
